File: examples/cpp/probe.cpp
#include "peripherals/peripherals.hpp"

#include <iostream>
#include <string>

namespace
{

bool ParseOptions(int argc, char **argv, std::string &serial_port)
{
    for (int i = 1; i < argc; ++i)
    {
        const std::string arg = argv[i];
        if (arg == "--serial-port" && i + 1 < argc)
        {
            serial_port = argv[++i];
        }
        else
        {
            std::cerr << "unknown argument: " << arg << '\n';
            return false;
        }
    }
    return true;
}

} // namespace

int main(int argc, char **argv)
{
    std::string serial_port = "/dev/ttyUSB0";
    if (!ParseOptions(argc, argv, serial_port))
    {
        return 1;
    }

    smrcore::peripherals::SpaceMouse spacemouse;
    smrcore::peripherals::SpaceMouseOptions sm_options;
    (void)spacemouse.Initialize(sm_options);
    auto sm_info = spacemouse.GetInfo();
    std::cout << "spacemouse model=" << sm_info.model
              << " rate_hz=" << sm_info.sample_rate_hz
              << " connected=" << sm_info.connected << '\n';

    smrcore::peripherals::FtSensor ft_sensor;
    smrcore::peripherals::FtSensorOptions ft_options;
    ft_options.serial_port = serial_port;
    (void)ft_sensor.Initialize(ft_options);
    auto ft_info = ft_sensor.GetInfo();
    std::cout << "ft_sensor model=" << ft_info.model
              << " rate_hz=" << ft_info.sample_rate_hz
              << " connected=" << ft_info.connected << '\n';
    return 0;
}

File: examples/python/probe.py
#!/usr/bin/env python3
"""Mirror of examples/cpp/probe.cpp."""

from __future__ import annotations

import argparse

from smrcore_peripherals import FtSensor, FtSensorOptions, SpaceMouse, SpaceMouseOptions


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Probe peripheral connectivity")
    parser.add_argument("--serial-port", default="/dev/ttyUSB0")
    return parser.parse_args()


def main() -> int:
    args = parse_args()

    spacemouse = SpaceMouse()
    sm_options = SpaceMouseOptions()
    spacemouse.Initialize(sm_options)
    sm_info = spacemouse.GetInfo()
    print(
        f"spacemouse model={sm_info.model} rate_hz={sm_info.sample_rate_hz} "
        f"connected={sm_info.connected}",
        flush=True,
    )

    ft_sensor = FtSensor()
    ft_options = FtSensorOptions(serial_port=args.serial_port)
    ft_sensor.Initialize(ft_options)
    ft_info = ft_sensor.GetInfo()
    print(
        f"ft_sensor model={ft_info.model} rate_hz={ft_info.sample_rate_hz} "
        f"connected={ft_info.connected}",
        flush=True,
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
